realm/src/main/java/io/realm/Realm.java

```diff
@@ -961,7 +961,38 @@ boolean contains(Class<? extends RealmObject> clazz) {
         }
 
         TableView tableView = table.getDistinctView(columnIndex);
-        return new RealmResults(this, tableView, clazz);
+        return new RealmResults<E>(this, tableView, clazz);
+    }
+
+    /**
+     * Return a distinct set of objects of a specific class. As a Realm is unordered, it is undefined which objects are
+     * returned in cases of multiple occurrences.
+     * This method is only available from a Looper thread.
+     *
+     * @param clazz the Class to get objects of.
+     * @param fieldName the field name.
+     * @return immediately an empty {@link RealmResults}. Users need to register a listener
+     * {@link io.realm.RealmResults#addChangeListener(RealmChangeListener)} to be notified
+     * when the query completes.
+     * @throws IllegalArgumentException if a field name does not exist or the field is not indexed.
+     */
+    public <E extends RealmObject> RealmResults<E> distinctAsync(Class<E> clazz, String fieldName) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("fieldName must be provided.");
+        }
+
+        Table table = this.getTable(clazz);
+        long columnIndex = table.getColumnIndex(fieldName);
+        if (columnIndex == -1) {
+            throw new IllegalArgumentException(String.format("Field name '%s' does not exist.", fieldName));
+        }
+
+        // check if the field is indexed
+        if (!table.hasSearchIndex(columnIndex)) {
```

Comment:
we discussed this yesterday in #java the point is to fail fast, I know the check is done in C++ but this will fail in a background thread, whereas in this approach we make sure all the arguments are valid